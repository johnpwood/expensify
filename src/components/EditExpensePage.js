import React from 'react';
import { connect } from 'react-redux'

import { editExpense, removeExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

export class EditExpensePage extends React.Component {
    onEdit = expense => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1>Edit Expense</h1>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onEdit} />
                <button
                    onClick={this.onRemove}>Delete</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: id => dispatch(removeExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

