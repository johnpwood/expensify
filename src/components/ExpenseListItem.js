import React from 'react'
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'

const ExpenseListItem = ({ id, description = '', amount = -1, createdAt=0 } ) => (
    <div>
        <Link to={`/edit/${id}`}><h3>{description}:</h3></Link>
        <p>Created at: {moment(createdAt).format('MMMM Do, YYYY')}</p>
        <p>amount: {numeral(amount/100).format('$0,0.00')}</p>
    </div>
)
ExpenseListItem.propTypes = {
    dispatch: PropTypes.func,
    id: PropTypes.string,
    description: PropTypes.string,
    amount: PropTypes.number,
    createdAt: PropTypes.number
}

export default ExpenseListItem;
