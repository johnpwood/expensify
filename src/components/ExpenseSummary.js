import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import totalExpenses from '../selectors/expenses-total'
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral'

export const ExpenseSummary = (props) => {
    const n = props.expenses.length
    return (
        <div>
            <p>Showing {n} expense{n === 1 ? '' : 's'} totalling {numeral(totalExpenses(props.expenses)/100).format('$0,0.00')}</p>
        </div>
    )
}

ExpenseSummary.propTypes = {
    expenses: PropTypes.array,
    filters: {

    }
}

const mapStateToProps = state => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseSummary)
