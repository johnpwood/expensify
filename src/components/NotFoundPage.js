import React from 'react';
import { Link } from 'react-router-dom'

const NotFoundPage = () => {
    return (
        <div>
            Page not found.
            <Link to="/">Go home.</Link>
        </div>
    )
}

export default NotFoundPage;