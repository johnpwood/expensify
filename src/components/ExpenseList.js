import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses.</p>
            ) : (
                props.expenses.map(x => (<ExpenseListItem {...x} key={x.id} />))
            )
        }
        
    </div>
)
ExpenseList.propTypes = {
    expenses: PropTypes.array,
    filters: {

    }
}

const mapStateToProps = state => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)
