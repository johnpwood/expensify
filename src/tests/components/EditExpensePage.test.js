import React from 'react'
import { shallow } from 'enzyme'

import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let wrapper, history, editExpense, removeExpense

beforeEach(() => { 
    history = { push: jest.fn() }
    editExpense = jest.fn()
    removeExpense = jest.fn()
    wrapper = shallow(<EditExpensePage
                            expense={expenses[2]}
                            history={history}
                            editExpense={editExpense}
                            removeExpense={removeExpense} />)
})

test('should render EditExpensePage', () => { 
    expect(wrapper).toMatchSnapshot()
})

test('should call appropriate function on form submission', () => {
    const expense = { ...expenses[2], id: undefined }
    wrapper.find('ExpenseForm').prop('onSubmit')(expense)
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expense)
})

test('should call appropriate function to remove the expense when the button is pressed', () => { 
    wrapper.find('button').simulate('click')
    expect(removeExpense).toHaveBeenCalledWith(expenses[2].id)
    expect(history.push).toHaveBeenCalledWith('/')
})