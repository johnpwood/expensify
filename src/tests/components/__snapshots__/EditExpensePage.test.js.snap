// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 1950,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "cbc123",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
